packages/aws-cdk/lib/api/hotswap-deployments.ts

```diff
@@ -48,18 +49,58 @@ function findAllHotswappableChanges(
   let foundNonHotswappableChange = false;
   stackChanges.resources.forEachDifference((logicalId: string, change: cfn_diff.ResourceDifference) => {
+    const nonHotswappableResourceFound = isNonHotswappableResourceChange(change);
     const lambdaFunctionShortCircuitChange = isHotswappableLambdaFunctionChange(logicalId, change, assetParamsWithEnv);
-    if (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) {
+    const stepFunctionShortCircuitChange = isHotswappableStepFunctionChange(logicalId, change, assetParamsWithEnv);
+
+    const requiresFullDeployment = (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) ||
+                                   (stepFunctionShortCircuitChange === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) ||
+                                   (nonHotswappableResourceFound === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT);
+
+    const irrelevant = (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.IRRELEVANT) &&
+                       (stepFunctionShortCircuitChange === ChangeHotswapImpact.IRRELEVANT) &&
+                       (nonHotswappableResourceFound === ChangeHotswapImpact.IRRELEVANT);
+
+    if (requiresFullDeployment) {
       foundNonHotswappableChange = true;
-    } else if (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.IRRELEVANT) {
+    } else if (irrelevant) {
       // empty 'if' just for flow-aware typing to kick in...
     } else {
-      hotswappableResources.push(lambdaFunctionShortCircuitChange);
+      if (typeof lambdaFunctionShortCircuitChange !== 'string') {
+        hotswappableResources.push(lambdaFunctionShortCircuitChange);
+      }
+
+      if (typeof stepFunctionShortCircuitChange !== 'string') {
+        hotswappableResources.push(stepFunctionShortCircuitChange);
+      }
     }
   });
   return foundNonHotswappableChange ? undefined : hotswappableResources;
 }
 
+/**
+ * returns `ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT` if a resource was deleted, or a change that we cannot short-circuit occured.
+ * Returns `ChangeHotswapImpact.IRRELEVANT` if a change that does not impact shortcircuiting occured, such as a metadata change.
+ */
+export function isNonHotswappableResourceChange(change: cfn_diff.ResourceDifference): ChangeHotswapImpact {
+  // change.newValue being undefined means the resource was removed; we can't short-circuit that change
+  if (!change.newValue) {
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+
+  const newResourceType = change.newValue.Type;
+  // Ignore Metadata changes
+  if (newResourceType === 'AWS::CDK::Metadata') {
+    return ChangeHotswapImpact.IRRELEVANT;
+  }
+  // The only other resource change we should see is a Lambda function or a StepFunctions StateMachine
+  if ((newResourceType !== 'AWS::Lambda::Function') && (newResourceType !== 'AWS::StepFunctions::StateMachine')) {
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
```

Comment:
This is a really bad idea. You're moving the knowledge of which resource types are supported from a local place like `isHotswappableLambdaFunctionChange()` to a global place here. This will not be maintainable in the long term.

Remove this condition from this function, and move it back to the specific "detector" functions.

Reply:
having 

```
  if (newResourceType !== 'AWS::Lambda::Function') {
    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
  }
```

inside of ` isHotswappableLambdaFunctionChange()` means that changes to non-lambda function resources require full deployments, ie changes to state machines require full deployments. We could check for each type in all of the detector functions, but then we have the same `if` block in each detector, which is why I extracted it to a separate function.

Reply:
Discussed offline - the problem mentioned will be solved by the algorithm described here: https://github.com/aws/aws-cdk/pull/16489#discussion_r708718668.